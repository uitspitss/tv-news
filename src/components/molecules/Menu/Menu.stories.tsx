import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import base from 'paths.macro';

import { Menu, MenuProps } from './Menu';

const path = base.replace(
  /^.+\/(atoms|molecules|organisms|templates|pages)\/(.+)\//,
  '$1/$2',
);

export default {
  title: `${path}`,
  component: Menu,
} as Meta;

const Template: Story<MenuProps> = (args) => <Menu {...args} />;

export const Default = Template.bind({});
Default.args = {
  trigger: <button type="button">trigger</button>,
  children: (
    <>
      <li>content</li>
      <li>content</li>
      <li>content</li>
    </>
  ),
};
