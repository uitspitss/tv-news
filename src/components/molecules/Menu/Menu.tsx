import React, { FC, memo, ReactNode } from 'react';
import Popup from 'reactjs-popup';
import { PopupPosition } from 'reactjs-popup/dist/types';
import tw, { styled } from 'twin.macro';

export type MenuProps = {
  /**
   * class for twin.macro
   */
  className?: string;
  /**
   * no children
   */
  children?: ReactNode;
  /**
   * trigger component
   */
  //  eslint-disable-next-line no-undef
  trigger: JSX.Element;
  /**
    "top left" | "top center" | "top right" | "right top" | "right center" | "right bottom" | "bottom left" | "bottom center" | "bottom right" | "left top" | "left center" | "left bottom" | "center center"
   */
  position?: PopupPosition;
};

type Props = MenuProps;

const Component: FC<Props> = (props) => {
  const { className, children, trigger, ...rest } = props;

  return (
    <Popup
      className={className}
      position="bottom left"
      trigger={trigger}
      arrow={false}
      {...rest}
    >
      <ul className="menuList">{children}</ul>
    </Popup>
  );
};

const StyledComponent = styled(Component)`
  &-content {
    .menuList {
      ${tw`rounded flex flex-col items-start`}

      & > li {
        ${tw`px-2 py-1`}
      }
    }
  }
`;

/**
 * リストメニュー
 */
export const Menu: FC<MenuProps> = memo((props) => {
  return <StyledComponent {...props} />;
});
