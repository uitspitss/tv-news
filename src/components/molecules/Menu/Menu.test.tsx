import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import { Menu } from './Menu';

describe('molecules/Menu', () => {
  test('snapshot test', () => {
    const { container } = render(
      <Menu trigger={<button type="button">trigger</button>}>
        <li>content</li>
        <li>content</li>
        <li>content</li>
      </Menu>,
    );
    expect(container).toMatchSnapshot();
  });

  describe('Behavior', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      render(
        <Menu trigger={<button type="button">trigger</button>}>
          <li>content</li>
          <li>content</li>
          <li>content</li>
        </Menu>,
      );
    });

    test('initial view', () => {
      expect(
        screen.getByRole('button', { name: 'trigger' }),
      ).toBeInTheDocument();
    });

    test('click, then show menu', () => {
      fireEvent.click(screen.getByRole('button', { name: 'trigger' }));

      expect(screen.getAllByText('content')).toHaveLength(3);
    });

    test('open menu, then click trigger, then close menu', () => {
      fireEvent.click(screen.getByRole('button', { name: 'trigger' }));

      expect(screen.getAllByText('content')).toHaveLength(3);

      fireEvent.click(screen.getByRole('button', { name: 'trigger' }));

      expect(screen.queryByText('content')).not.toBeInTheDocument();
    });
  });
});
