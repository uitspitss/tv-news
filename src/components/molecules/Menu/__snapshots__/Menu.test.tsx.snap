// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecules/Menu snapshot test 1`] = `
<div>
  <button
    aria-describedby="popup-1"
    type="button"
  >
    trigger
  </button>
</div>
`;
